### PR TITLE
[`Flash Attention 2`] Add flash attention 2 for GPT-Neo-X

### LINKED ISSUE #26444: add flush attention support model
I want to use flush attention with Neox.
When I utilize it?

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

Adds flash attention support for GPT-Neo-X

Fixes: https://github.com/huggingface/transformers/issues/26444

### CODE REVIEW

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py [hunk 1]
```diff
@@ -281,6 +302,204 @@ def _attn(self, query, key, value, attention_mask=None, head_mask=None):
...
+            logger.warning_once(
+                "The input hidden states seems to be silently casted in float32, this might be related to"
+                " the fact you have upcasted embedding or layer norm layers in float32. We will cast back the input in"
+                " float16."
+            )
+
+            query = query.to(torch.float16)
+            key = key.to(torch.float16)
+            value = value.to(torch.float16)
+
```
Comment (ArthurZucker) [inline src/transformers/models/gpt_neox/modeling_gpt_neox.py:393]: We should take into account `bfloat16` here as well

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py [hunk 2]
```diff
@@ -281,6 +302,204 @@ def _attn(self, query, key, value, attention_mask=None, head_mask=None):
...
+        # Permute to get the expected shape for Flash Attention
+        query = query.permute(0, 2, 1, 3)
+        key = key.permute(0, 2, 1, 3)
+        value = value.permute(0, 2, 1, 3)
+
+        # In PEFT, usually we cast the layer norms in float32 for training stability reasons
+        # therefore the input hidden states gets silently casted in float32. Hence, we need
+        # cast them back in float16 just to be sure everything works as expected.
+        # This might slowdown training & inference so it is recommended to not cast the LayerNorms
+        # in fp32. (LlamaRMSNorm handles it correctly)
```
Comment (ArthurZucker) [inline src/transformers/models/gpt_neox/modeling_gpt_neox.py:381]: Is this also true for GPTNeoX? (Comment is the same as Llama 😓 )

**Review by ArthurZucker — APPROVED** [review]
LGTM but left a few nits

**Comment by btrude** [comment]
Any plans on completing this or should someone else pick it up? For what it's worth, this implementation is working very well for me 👍

**Comment by younesbelkada** [comment]
cc @amyeroberts let me know if I need to address anything else in this PR!

**Comment by avnermay** [comment]
Checking on the progress here. What's the ETA on merging this with the main branch? Thanks!

File: docs/source/en/model_doc/gpt_neox.md [hunk 3]
```diff
@@ -61,6 +61,31 @@ The `generate()` method can be used to generate text using GPT Neo model.
...
+
+To load a model using Flash Attention 2, we can pass the `use_flash_attention_2` flag to [`.from_pretrained`](https://huggingface.co/docs/transformers/main/en/main_classes/model#transformers.PreTrainedModel.from_pretrained). We'll also load the model in half-precision (e.g. `torch.float16`), since it results in almost no degradation to audio quality but significantly lower memory usage and faster inference:
+
+```python
+>>> from transformers import GPTNeoXForCausalLM, GPTNeoXTokenizerFast
+
+model = GPTNeoXForCausalLM.from_pretrained("EleutherAI/gpt-neox-20b", torch_dtype=torch.float16, use_flash_attention_2=True).to(device)
+...
+```
+
```
Comment (amyeroberts) [inline docs/source/en/model_doc/gpt_neox.md:88]: There should be a comparison on performance with and without FA here

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py [hunk 4]
```diff
@@ -277,6 +299,211 @@ def _attn(self, query, key, value, attention_mask=None, head_mask=None):
...
+
+            attn_output = pad_input(attn_output_unpad, indices_q, batch_size, query_length)
+        else:
+            attn_output = flash_attn_func(
+                query_states, key_states, value_states, dropout, softmax_scale=softmax_scale, causal=self.is_causal
+            )
+
+        return attn_output
+
+    def _upad_input(self, query_layer, key_layer, value_layer, attention_mask, query_length):
```
Comment (amyeroberts) [inline src/transformers/models/gpt_neox/modeling_gpt_neox.py:468]: Is this different from the _upad_input in llama? If not - could we add a `# Copied from` header here?

**Review by amyeroberts — APPROVED** [review]
LGTM - thanks for adding! 

Just needs a performance example to be added to the docs before merging

File: src/transformers/models/gpt_neox/modeling_gpt_neox.py (hunk 4 above)
Comment (younesbelkada) [reply src/transformers/models/gpt_neox/modeling_gpt_neox.py:483]: Makes sense, done!

File: docs/source/en/model_doc/gpt_neox.md [hunk 5]
```diff
@@ -61,6 +61,40 @@ The `generate()` method can be used to generate text using GPT Neo model.
...
+
+### Expected speedups
+
+Below is an expected speedup diagram that compares pure inference time between the native implementation in transformers using `stockmark/gpt-neox-japanese-1.4b` checkpoint and the Flash Attention 2 version of the model using two different sequence lengths.
```
Comment (younesbelkada) [inline docs/source/en/model_doc/gpt_neox.md:92]:
```suggestion
Below is an expected speedup diagram that compares pure inference time between the native implementation in transformers using `stockmark/gpt-neox-japanese-1.4b` checkpoint and the Flash Attention 2 version of the model using a sequence length of 2048.
```

**Pull request merged**